File: frontend/src/lib/types/incident.ts
import type { User } from './user';
import type { Alert } from './alert';

export type IncidentSeverity = 'critical' | 'high' | 'medium' | 'low';
export type IncidentStatus = 'investigating' | 'identified' | 'monitoring' | 'resolved';
export type AlertPriority = 'P1' | 'P2' | 'P3' | 'P4' | 'P5';
export type ResponderRole = 'incident_commander' | 'responder';
export type TimelineEventType =
	| 'created'
	| 'status_changed'
	| 'severity_changed'
	| 'responder_added'
	| 'responder_removed'
	| 'note_added'
	| 'alert_linked'
	| 'alert_unlinked'
	| 'resolved';

export interface Incident {
	id: string;
	organization_id: string;
	title: string;
	description?: string;
	severity: IncidentSeverity;
	status: IncidentStatus;
	priority: AlertPriority;
	created_by_user_id: string;
	assigned_to_team_id?: string;
	started_at: string;
	resolved_at?: string;
	created_at: string;
	updated_at: string;
}

export interface IncidentResponder {
	id: string;
	incident_id: string;
	user_id: string;
	role: ResponderRole;
	added_at: string;
}

export interface ResponderWithUser extends IncidentResponder {
	user: User;
}

export interface IncidentTimelineEvent {
	id: string;
	incident_id: string;
	event_type: TimelineEventType;
	user_id?: string;
	description: string;
	metadata: Record<string, any>;
	created_at: string;
}

export interface TimelineEventWithUser extends IncidentTimelineEvent {
	user?: User;
}

export interface IncidentAlert {
	id: string;
	incident_id: string;
	alert_id: string;
	linked_at: string;
	linked_by_user_id?: string;
}

export interface IncidentAlertWithDetails extends IncidentAlert {
	alert: Alert;
}

export interface IncidentWithDetails extends Incident {
	responders?: ResponderWithUser[];
	alerts?: IncidentAlertWithDetails[];
	timeline?: TimelineEventWithUser[];
}

// Request types
export interface CreateIncidentRequest {
	title: string;
	description?: string;
	severity: IncidentSeverity;
	priority: AlertPriority;
	assigned_to_team_id?: string;
}

export interface UpdateIncidentRequest {
	title?: string;
	description?: string;
	severity?: IncidentSeverity;
	status?: IncidentStatus;
	priority?: AlertPriority;
	assigned_to_team_id?: string;
}

export interface AddResponderRequest {
	user_id: string;
	role: ResponderRole;
}

export interface UpdateResponderRoleRequest {
	role: ResponderRole;
}

export interface AddNoteRequest {
	note: string;
}

export interface LinkAlertRequest {
	alert_id: string;
}

export interface ListIncidentsParams {
	status?: IncidentStatus[];
	severity?: IncidentSeverity[];
	assigned_to_team_id?: string;
	search?: string;
	page?: number;
	page_size?: number;
}

export interface ListIncidentsResponse {
	incidents: Incident[];
	total: number;
	page: number;
	page_size: number;
}

File: backend/internal/domain/incident.go
package domain

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

// IncidentSeverity represents the severity level of an incident
type IncidentSeverity string

const (
	IncidentSeverityCritical IncidentSeverity = "critical"
	IncidentSeverityHigh     IncidentSeverity = "high"
	IncidentSeverityMedium   IncidentSeverity = "medium"
	IncidentSeverityLow      IncidentSeverity = "low"
)

// IsValid checks if the severity is valid
func (s IncidentSeverity) IsValid() bool {
	switch s {
	case IncidentSeverityCritical, IncidentSeverityHigh, IncidentSeverityMedium, IncidentSeverityLow:
		return true
	}
	return false
}

func (s IncidentSeverity) String() string {
	return string(s)
}

// IncidentStatus represents the current status of an incident
type IncidentStatus string

const (
	IncidentStatusInvestigating IncidentStatus = "investigating"
	IncidentStatusIdentified    IncidentStatus = "identified"
	IncidentStatusMonitoring    IncidentStatus = "monitoring"
	IncidentStatusResolved      IncidentStatus = "resolved"
)

// IsValid checks if the status is valid
func (s IncidentStatus) IsValid() bool {
	switch s {
	case IncidentStatusInvestigating, IncidentStatusIdentified, IncidentStatusMonitoring, IncidentStatusResolved:
		return true
	}
	return false
}

func (s IncidentStatus) String() string {
	return string(s)
}

// Incident represents an incident
type Incident struct {
	ID               uuid.UUID
	OrganizationID   uuid.UUID
	Title            string
	Description      *string
	Severity         IncidentSeverity
	Status           IncidentStatus
	Priority         AlertPriority
	CreatedByUserID  uuid.UUID
	AssignedToTeamID *uuid.UUID
	StartedAt        time.Time
	ResolvedAt       *time.Time
	CreatedAt        time.Time
	UpdatedAt        time.Time
}

// ResponderRole represents the role of an incident responder
type ResponderRole string

const (
	ResponderRoleIncidentCommander ResponderRole = "incident_commander"
	ResponderRoleResponder         ResponderRole = "responder"
)

// IsValid checks if the responder role is valid
func (r ResponderRole) IsValid() bool {
	switch r {
	case ResponderRoleIncidentCommander, ResponderRoleResponder:
		return true
	}
	return false
}

func (r ResponderRole) String() string {
	return string(r)
}

// IncidentResponder represents a user assigned to an incident
type IncidentResponder struct {
	ID         uuid.UUID
	IncidentID uuid.UUID
	UserID     uuid.UUID
	Role       ResponderRole
	AddedAt    time.Time
}

// ResponderWithUser extends IncidentResponder with user details
type ResponderWithUser struct {
	IncidentResponder
	User *User
}

// TimelineEventType represents the type of timeline event
type TimelineEventType string

const (
	TimelineEventCreated          TimelineEventType = "created"
	TimelineEventStatusChanged    TimelineEventType = "status_changed"
	TimelineEventSeverityChanged  TimelineEventType = "severity_changed"
	TimelineEventResponderAdded   TimelineEventType = "responder_added"
	TimelineEventResponderRemoved TimelineEventType = "responder_removed"
	TimelineEventNoteAdded        TimelineEventType = "note_added"
	TimelineEventAlertLinked      TimelineEventType = "alert_linked"
	TimelineEventAlertUnlinked    TimelineEventType = "alert_unlinked"
	TimelineEventResolved         TimelineEventType = "resolved"
)

// IsValid checks if the timeline event type is valid
func (t TimelineEventType) IsValid() bool {
	switch t {
	case TimelineEventCreated, TimelineEventStatusChanged, TimelineEventSeverityChanged,
		TimelineEventResponderAdded, TimelineEventResponderRemoved, TimelineEventNoteAdded,
		TimelineEventAlertLinked, TimelineEventAlertUnlinked, TimelineEventResolved:
		return true
	}
	return false
}

func (t TimelineEventType) String() string {
	return string(t)
}

// IncidentTimelineEvent represents an event in the incident timeline
type IncidentTimelineEvent struct {
	ID          uuid.UUID
	IncidentID  uuid.UUID
	EventType   TimelineEventType
	UserID      *uuid.UUID
	Description string
	Metadata    map[string]interface{}
	CreatedAt   time.Time
}

// TimelineEventWithUser extends IncidentTimelineEvent with user details
type TimelineEventWithUser struct {
	IncidentTimelineEvent
	User *User
}

// IncidentAlert represents a link between an incident and an alert
type IncidentAlert struct {
	ID             uuid.UUID
	IncidentID     uuid.UUID
	AlertID        uuid.UUID
	LinkedAt       time.Time
	LinkedByUserID *uuid.UUID
}

// IncidentAlertWithDetails extends IncidentAlert with alert details
type IncidentAlertWithDetails struct {
	IncidentAlert
	Alert *Alert
}

// IncidentFilter represents filters for listing incidents
type IncidentFilter struct {
	OrganizationID   uuid.UUID
	Status           []IncidentStatus
	Severity         []IncidentSeverity
	AssignedToTeamID *uuid.UUID
	Search           *string
	Limit            int
	Offset           int
}

// Validate validates the incident filter
func (f *IncidentFilter) Validate() error {
	if f.Limit < 0 {
		return fmt.Errorf("limit must be non-negative")
	}
	if f.Offset < 0 {
		return fmt.Errorf("offset must be non-negative")
	}
	return nil
}

// IncidentWithDetails extends Incident with related data
type IncidentWithDetails struct {
	Incident
	Responders []*ResponderWithUser
	Alerts     []*IncidentAlertWithDetails
	Timeline   []*TimelineEventWithUser
}
